spec/abci++/abci++_basic_concepts_002_draft.md

```diff
@@ -106,95 +156,104 @@ the `LoadSnapshotChunk` method.
+exectution. An Application implementing immediate exectution may additionally wish to store
+certain data resulting from the block's execution in the same block that has just been executed.
+This brings about a new execution model, called *same-block execution*. An Application
+implementing this execution model, upon receiving a raw proposal via `RequestPrepareProposal`
+and potentially modifying its transaction list, fully executes the resulting prepared proposal
+as though it was the decided block (immediate execution), and
+the results of the block execution are used as follows:
+
+- The block execution may generate a set of events. The Application should store these events and
+  return them back to Tendermint during the `FinalizeBlock` call if the block is finally decided.
+- The Merkle root resulting from executing the prepared proposal is provided in
   `ResponsePrepareProposal` and thus refers to the **current block**. Tendermint
   will use it in the prepared proposal's header.
-* likewise, the transaction results from executing the prepared proposal are
+- Likewise, the transaction results from executing the prepared proposal are
   provided in `ResponsePrepareProposal` and refer to the transactions in the
   **current block**. Tendermint will use them to calculate the results hash
   in the prepared proposal's header.
-* The consensus parameter updates and validator updates are also provided in
+- The consensus parameter updates and validator updates are also provided in
   `ResponsePrepareProposal` and reflect the result of the prepared proposal's
   execution. They come into force in height H+1 (as opposed to the H+2 rule
   in next-block execution model).
 
 If the Application decides to keep the next-block execution model, it will not
-provide any data in `ResponsePrepareProposal`, other than an optionally modified
-transaction list.
+provide any data in `ResponsePrepareProposal`, other than a potentially modified
+transaction list. The Application may choose to perform immediate execution even in next-block
+execution mode, however same-block execution mode *requires* immediate execution.
 
 In the long term, the execution model will be set in a new boolean parameter
 *same_block* in `ConsensusParams`.
 It **must not** be changed once the blockchain has started unless the Application
-developers _really_ know what they are doing.
+developers *really* know what they are doing.
 However, modifying `ConsensusParams` structure cannot be done lightly if we are to
 preserve blockchain compatibility. Therefore we need an interim solution until
-soft upgrades are specified and implemented in Tendermint. This somewhat _unsafe_
+soft upgrades are specified and implemented in Tendermint. This somewhat *unsafe*
 solution consists in Tendermint assuming same-block execution if the Application
 fills the above mentioned fields in `ResponsePrepareProposal`.
 
-### Tendermint timeouts in same-block execution
+## Tendermint proposal timeout
```

Comment:
```suggestion
### Tendermint proposal timeout
```
I think this is still part of the same block next block execution piece right?

Reply:
I separated it for two reasons:

* The proposal timeout problem is not really linked to same-block execution, but to _immediate_ execution. You can have immediate execution (i.e., fully executing the block at `PrepareProposal`/`ProcessProposal` time) even if same-block execution (i.e., actually saving app_hash, tx_results, etc., in the block to be proposaed) is not implemented.
IOW, same-block does need immediate, but immediate does not need same-block.
* So I separated this in preparation for removing the previous "Next-block vs. same-block" in the v0.36.x branch.